From 46bded89e9e2784a894f744eba9d8bd02b0d6d3f Mon Sep 17 00:00:00 2001
From: Junxiao Chang <junxiao.chang@intel.com>
Date: Fri, 22 Feb 2019 17:26:48 +0800
Subject: [PATCH 2313/2367] uart: fix uart locking error

This error is introduced by merge conflict in 4.14.97(190205T003243Z).
One change is "5df3737 serial: core: add support of runtime PM".
Another is "be9497a uart: Fix crash in uart_write and uart_put_char".

This change is to fix conflict error between these two commits.

Signed-off-by: Junxiao Chang <junxiao.chang@intel.com>
---
 drivers/tty/serial/serial_core.c | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/drivers/tty/serial/serial_core.c b/drivers/tty/serial/serial_core.c
index fff7370..6f55f75 100644
--- a/drivers/tty/serial/serial_core.c
+++ b/drivers/tty/serial/serial_core.c
@@ -585,17 +585,17 @@ static int uart_put_char(struct tty_struct *tty, unsigned char c)
 	int ret = 0;
 
 	circ = &state->xmit;
-	port = uart_port_lock(state, flags);
-	if (!circ->buf) {
-		uart_port_unlock(port, flags);
-		return 0;
-	}
-
 	port = uart_port_ref_no_rpm(state);
 	if (!port)
 		return 0;
 
 	spin_lock_irqsave(&port->lock, flags);
+	if (!circ->buf) {
+		spin_unlock_irqrestore(&port->lock, flags);
+		uart_port_deref_no_rpm(port);
+		return 0;
+	}
+
 	if (uart_circ_chars_free(circ) != 0) {
 		circ->buf[circ->head] = c;
 		circ->head = (circ->head + 1) & (UART_XMIT_SIZE - 1);
@@ -629,18 +629,18 @@ static int uart_write(struct tty_struct *tty,
 		return -EL3HLT;
 	}
 
-	port = uart_port_lock(state, flags);
-	circ = &state->xmit;
-	if (!circ->buf) {
-		uart_port_unlock(port, flags);
-		return 0;
-	}
-
 	port = uart_port_ref_no_rpm(state);
 	if (!port)
 		return 0;
 
 	spin_lock_irqsave(&port->lock, flags);
+	circ = &state->xmit;
+	if (!circ->buf) {
+		spin_unlock_irqrestore(&port->lock, flags);
+		uart_port_deref_no_rpm(port);
+		return 0;
+	}
+
 	while (port) {
 		c = CIRC_SPACE_TO_END(circ->head, circ->tail, UART_XMIT_SIZE);
 		if (count < c)
-- 
2.7.4

